From 54ae996bc21918c346aafc7b8eed6ef3df0a1977 Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Tue, 17 Feb 2026 23:02:24 +0000
Subject: [PATCH 06/11] Allow fingerprint HAL to read serialno_prop

GrapheneOS restricts access to serialno_prop as a privacy hardening
measure. However, many vendor fingerprint HALs (including Agold/
Chipsailing on 8849 and Unihertz devices) read the device serial
number to bind fingerprint templates to the device. Without this
access, the HAL cannot initialize properly and fingerprint enrollment
fails silently.

Add hal_fingerprint_server to the neverallow exception list and grant
it get_prop access to serialno_prop.
---
 private/domain.te          | 1 +
 private/hal_fingerprint.te | 4 ++++
 2 files changed, 5 insertions(+)

diff --git a/private/domain.te b/private/domain.te
index 7d5869e..9a65d0f 100644
--- a/private/domain.te
+++ b/private/domain.te
@@ -884,6 +884,7 @@ neverallow {
   -hal_camera_server
   -hal_cas_server
   -hal_drm_server
+  -hal_fingerprint_server
   -hal_keymint_server
   userdebug_or_eng(`-incidentd')
   -init
diff --git a/private/hal_fingerprint.te b/private/hal_fingerprint.te
index 07a0a65..861f62f 100644
--- a/private/hal_fingerprint.te
+++ b/private/hal_fingerprint.te
@@ -7,6 +7,10 @@ hal_attribute_service(hal_fingerprint, hal_fingerprint_service)
 
 binder_use(hal_fingerprint_server)
 
+# Many vendor fingerprint HALs read the device serial number to bind
+# fingerprint templates to the device via the TEE.
+get_prop(hal_fingerprint_server, serialno_prop)
+
 allow { hal_fingerprint -coredomain } fingerprint_vendor_data_file:file { create_file_perms };
 allow { hal_fingerprint -coredomain } fingerprint_vendor_data_file:dir rw_dir_perms;
 
-- 
2.34.1

